File: lib/sensetime_api.rb
require 'rest-client'

module SensetimeAPI
  def self.selfie_idnumber_verification file, id, name
    url = 'https://v1-auth-api.visioncloudapi.com/identity/selfie_idnumber_verification'
    params = {
      selfie_file: file,
      id_number: id,
      name: name
    }

    send_request url, params
  end

  def self.historical_selfie_verification image, file
    url = 'https://v1-auth-api.visioncloudapi.com/identity/historical_selfie_verification'
    params = {
      selfie_file: file,
      historical_selfie_file: Paperclip.io_adapters.for(image.file)
    }

    send_request url, params
  end

  def self.send_request url, params
    params.merge!(api_id: ENV["SENSETIME_API_ID"], api_secret: ENV["SENSETIME_API_SECRET"])
    Rails.logger.debug "  Request #{url} with #{params}."
    result = RestClient.post(url, params)
    Rails.logger.debug "  Response: #{result}"
    JSON.parse result
  rescue => error
    Rails.logger.error error.response.body
    raise error
  end
end

File: web/helpers/sensetime_api.ex
defmodule Scanner.SensetimeAPI do
  require Logger

  def selfie_idnumber_verification file, id, name do
    [api_id: api_id, api_secret: api_secret] = Application.get_env(:scanner, :sensetime)
    url = "https://v1-auth-api.visioncloudapi.com/identity/selfie_idnumber_verification"
    body = {:multipart, [
      {:file, file.path, {["form-data"], [{"name", "selfie_file"}, {"filename", file.filename}]}, [{"Content-Type", file.content_type}]},
      {"api_id", api_id},
      {"api_secret", api_secret},
      {"id_number", id},
      {"name", name},
    ]}

    Logger.debug "Request #{url} with #{inspect(body)}"

    result =
      case HTTPoison.post(url, body) do
        {:ok, response} -> Poison.decode response.body
        {:error, reason} -> {:error, reason}
      end

    Logger.debug "Response #{inspect(result)}"

    result
  end
end
